File: dart/bob/lib/bob.dart
class Bob {
  String response(String arg) {
    final isQuestion = arg.trim().endsWith('?');
    final containsCharacter = RegExp(r'[a-zA-Z]+').hasMatch(arg);
    final isUpper = arg.toUpperCase() == arg;
    if (isQuestion) {
      if (isUpper && containsCharacter) {
        return 'Calm down, I know what I\'m doing!';
      }
      return 'Sure.';
    }
    if (isUpper && containsCharacter) return 'Whoa, chill out!';
    if (arg.trim() == '') return 'Fine. Be that way!';
    return 'Whatever.';
  }
}

File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"strings"
	"unicode"
)

// Remark is a convenience type for identifying what kind of remark it is.
type Remark struct {
	remark string
}

func (r Remark) isSilence() bool {
	return r.remark == ""
}

func (r Remark) isShouting() bool {
	hasLetters := strings.IndexFunc(r.remark, unicode.IsLetter) >= 0
	isUpcased := strings.ToUpper(r.remark) == r.remark

	return hasLetters && isUpcased
}

func (r Remark) isQuestion() bool {
	return strings.HasSuffix(r.remark, "?")
}

func (r Remark) isExasperated() bool {
	return r.isShouting() && r.isQuestion()
}

func newRemark(remark string) Remark {
	return Remark{strings.TrimSpace(remark)}
}

// Hey should have a comment documenting it.
func Hey(remark string) string {
	r := newRemark(remark)
	switch {
	case r.isSilence():
		return "Fine. Be that way!"
	case r.isExasperated():
		return "Calm down, I know what I'm doing!"
	case r.isShouting():
		return "Whoa, chill out!"
	case r.isQuestion():
		return "Sure."
	default:
		return "Whatever."
	}
}
